encoding/rlp/rlp.go

```diff
@@ -0,0 +1,342 @@
+	if len(length)%2 == 1 {
+		length = fmt.Sprintf("0%s", length)
+	}
+
+	dec, err := hex.DecodeString(length)
+	if err != nil {
+		return nil
+	}
+
+	return append([]byte{uint8(0xb7 + len(string(n)))}, dec...)
+}
+
+func encodeString(input []byte) ([]byte, error) {
+	if len(input) == 0 {
+		return []byte{EmptyString}, nil
+	} else if len(input) == 1 {
+		return encodeUint8(input[0])
+	} else {
+		return append(encodeLength(len(input), EmptyString), []byte(input)...), nil
+	}
+}
+
+func encodeList(val reflect.Value) ([]byte, error) {
+	if val.Len() == 0 {
+		return []byte{EmptySlice}, nil
+	}
+
+	out := make([][]byte, 0)
+	for i := 0; i < val.Len(); i++ {
+		data, err := encode(val.Index(i).Interface())
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, data)
+	}
+
+	sum := bytes.Join(out, []byte{})
+	return append(encodeLength(len(sum), EmptySlice), sum...), nil
+}
+
+func encodeStruct(val reflect.Value) ([]byte, error) {
+	out := make([][]byte, 0)
+
+	for i := 0; i < val.NumField(); i++ {
+		data, err := encode(val.Field(i).Interface())
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, data)
+	}
+	sum := bytes.Join(out, []byte{})
+	return append(encodeLength(len(sum), EmptySlice), sum...), nil
+}
+
+func encode(input interface{}) ([]byte, error) {
+	val := reflect.ValueOf(input)
+	typ := reflect.TypeOf(input)
+
+	switch val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		i := val.Int()
+		if i < 0 {
+			return nil, fmt.Errorf("cannot rlp encode negative integer")
+		}
+		return encodeUint64(uint64(i))
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return encodeUint64(val.Uint())
+	case reflect.Bool:
+		if val.Bool() {
+			return []byte{0x01}, nil
+		}
+		return []byte{EmptyString}, nil
+	case reflect.String:
+		return encodeString([]byte(reflect.ValueOf(input).String()))
+	case reflect.Slice:
+		switch typ.Elem().Kind() {
+		case reflect.Uint8:
+			return encodeString(reflect.ValueOf(input).Bytes())
+		default:
+			return encodeList(val)
+		}
+	case reflect.Struct:
+		return encodeStruct(val)
+	default:
+		return []byte{EmptyString}, nil
+	}
+}
+
+func Encode(input interface{}) ([]byte, error) {
+	return encode(input)
+}
+
+type fields struct {
+	fields [][]byte
+}
+
+func (f *fields) add(element []byte) {
+	f.fields = append(f.fields, element)
+}
+
+func decode(in []byte, out *fields) error {
+	if len(in) == 0 {
+		return nil
+	}
+
+	offset, length, typ, err := decodeLength(in)
+	if err != nil {
+		return err
+	}
+
+	switch typ {
+	case reflect.String:
+		out.add(in[offset:length])
+	case reflect.Slice:
+		err = decode(in[offset:length], out)
+		if err != nil {
+			return err
+		}
+	}
+
+	return decode(in[length:], out)
+}
+
+func decodeLength(input []byte) (uint8, uint8, reflect.Kind, error) {
+	length := uint8(len(input))
+
+	if length == 0 {
+		return 0, 0, reflect.Invalid, ErrNoInput
+	}
+
+	prefix := uint8(input[0])
```

Comment:
uint8 and byte should be implicitly convertible, do we need explicity cast here?